Documentation/WorkflowSwiftUI Adoption Guide.md

```diff
@@ -0,0 +1,448 @@
+        context.makeStateAccessor(state: state)
+    }
+}
+
+```
+
+For simple workflows with a single action, you can generate a model from your state and action:
+
+```swift
+struct SingleActionWorkflow: Workflow {
+    typealias Output = Never
+
+    @ObservableState
+    struct State {
+        var counter = 0
+    }
+
+    enum Action: WorkflowAction {
+        typealias WorkflowType = SingleActionWorkflow
+        case increment
+
+        func apply(toState state: inout State) -> Never? {
+            state.counter += 1
+            return nil
+        }
+    }
+
+    func makeInitialState() -> State {
+        .init()
+    }
+
+    func render(
+        state: State,
+        context: RenderContext<Self>
+    ) -> ActionModel<State, Action> {
+        context.makeActionModel(state: state)
+    }
+}
+
+```
+
+For complex workflows that have multiple actions or compose observable models from child
+workflows, you can create a custom model that conforms to `ObservableModel`:
+
+```swift
+struct ComplexWorkflow: Workflow {
+    typealias Output = Never
+
+    @ObservableState
+    struct State {
+        var counter = 0
+    }
+
+    enum UpAction: WorkflowAction {
+        typealias WorkflowType = ComplexWorkflow
+        case increment
+
+        func apply(toState state: inout State) -> Never? {
+            state.counter += 1
+            return nil
+        }
+    }
+
+    enum DownAction: WorkflowAction {
+        typealias WorkflowType = ComplexWorkflow
+        case decrement
+
+        func apply(toState state: inout State) -> Never? {
+            state.counter -= 1
+            return nil
+        }
+    }
+
+    func makeInitialState() -> State {
+        .init()
+    }
+
+    func render(
+        state: State,
+        context: RenderContext<Self>
+    ) -> CustomModel {
+        CustomModel(
+            accessor: context.makeStateAccessor(state: state),
+            child: TrivialWorkflow().rendered(in: context),
+            up: context.makeSink(of: UpAction.self),
+            down: context.makeSink(of: DownAction.self)
+        )
+    }
+}
+
+struct CustomModel: ObservableModel {
+    var accessor: StateAccessor<ComplexWorkflow.State>
+
+    var child: TrivialWorkflow.Rendering
+
+    var up: Sink<ComplexWorkflow.UpAction>
+    var down: Sink<ComplexWorkflow.DownAction>
+}
+```
+
+## Views
+
+In your View, you’ll access state and sinks via a property of type `Store<Model>`. This type wraps the model that your workflow renders, and provides conveniences for access the state, sinks, and child stores.
+
+If targeting iOS 16 or below, you’ll need to wrap your view’s body in `WithPerceptionTracking`. This is a component of the `Perception` backport that allows observation to work on iOS before iOS 17.
+
+```swift
+struct PersonState {
+  var name: String
+  var age: Int
+}
+
+enum PersonAction {
+  case giveHighFive
+}
+
+typealias PersonModel = ActionModel<PersonState, PersonAction>
+
+struct PersonView: View {
```

Comment:
Spitballing on how to avoid people needing `WithPerceptionTracking`:

We define a `ObservedView` protocol,

```swift
protocol ObservingView: View {

   var observing : some View { get }

}

extension ObservingView {
   var body : some View {
      WithPerceptionTracking { observing }
   }
}
```

Where folks would implement `observing`

Which requires learning the type, but once you know, you won't need the perception tracking wrapper, I think?

Reply:
@watt Clients will be able to drop `WithPerceptionTracking` once the deployment target is bumped to 17, right?

If that's the case, I probably wouldn't introduce a new wrapper protocol _just_ for that. Maybe we'll end up wanting/needing to wrap up some other functionality/convenience at the Market level and we can figure out how best to slip it in there.

Reply:
> @watt Clients will be able to drop `WithPerceptionTracking` once the deployment target is bumped to 17, right?

That's right. You also need to use it inline in some lazy contexts, like `GeometryReader`.